src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeOperator.scala

```diff
@@ -224,59 +235,92 @@ class TezosNodeOperator(val node: TezosRPCInterface)(implicit executionContext:
+      (network, hash, maybeOffset) => futureToIO(getBlock(network, hash, maybeOffset))
+
+    def blockExists: Block => IO[Boolean] =
+      block => {
+        val exists = TezosDatabaseOperations.blockExists(block.metadata.hash)
+        futureToIO(dbHandle.run(exists))
+      }
+
+    def blockHasBeenInvalidated: Block => IO[Boolean] =
+      block => {
+        val invalidated = TezosDatabaseOperations.blockExistsInInvalidatedBlocks(block.metadata.hash)
+        futureToIO(dbHandle.run(invalidated))
       }
 
-    val foldFunction : (Future[List[Block]], Future[Block]) => Future[List[Block]] =
-      (acc, futureBlock) => {
-        val p = predicate(futureBlock)
-        p match {
-          case (false, false) =>
-            futureBlock.flatMap{ block =>
-              val action = Tables.Blocks.filter(_.level === block.metadata.header.level)
-              //val blocks = dbHandle.run(action)
-              //TezosDatabaseOperations.writeInvalidatedBlocksIO(blocks)
-              for {
-                a <- acc
-              } yield  block :: a
-            }
-          case (true, true) => {
-            futureBlock.flatMap { block =>
-              //val (action1, action2) = TezosDatabaseOperations.updateInvalidatedBlockIO(block)
-              //db.run(action1)
-              //db.run(action2)
-              acc
-            }
-          }
+    def predicate: Block => IO[(Boolean, Boolean)] =
+      block => Applicative[IO].product(blockExists(block), blockHasBeenInvalidated(block))
+
+    /*
+    Given an offset from the block where the hash was detected, figure out if we need to collect that
+    block or perform a write action to our database.
+     */
+    def unfoldingFunction(offset: Int): IO[Option[(BlockWithAction, Int)]] = {
+
+      //rename to make more clear
+      def f(exists: Boolean,
+            invalidated: Boolean,
+            block: Block,
+            offset: Int): Option[(BlockWithAction, Int)] = {
+        if (exists && !invalidated) None
+          //add logging, because second case should be impossible, you should not have a block that does not exists
+          //being invalidated
+        else if (!exists && invalidated) None
+        else if (exists && invalidated) {
+          val blockWithAction = BlockWithAction(block, RevalidateBlock)
+          Some(blockWithAction, offset + 1)
+        }
+        else {
+          val blockWithAction = BlockWithAction(block, WriteAndInvalidateBlock)
+          Some(blockWithAction, offset + 1)
         }
       }
 
-    offsets
-      .map{ offset => {getBlock(network, hash, Some(offset))}}
-      .takeWhile{ block => predicate(block) != (true, false)}
-      .foldLeft(Future.successful(List.empty[Block]))(foldFunction)
+      for {
+        block <- getBlockIO(network, hash, Some(offset))
+        tuple <- predicate(block)
+        //scala compiler complained about tuples not having a withFilter instance with IO
```

Comment:
I had a chance to verify that this is related to binding with pattern-matching: the compiler rewrites that as `withFilter` calls to check that the pattern works

[further reference](https://github.com/oleg-py/better-monadic-for#desugaring-for-patterns-without-withfilters)